### PR TITLE
event_logs, telemetry_export: do not interrupt insert on context cancel

### PR DESCRIPTION
**Pull request by bobheadxi**
When a request context is cancelled, the [corresponding event database insert can fail on context cancellation](https://sourcegraph.slack.com/archives/C03V51C8K53/p1697072391851359) - but this is highly unlikely to be desirable behaviour, since we probably want to proceed with persisting the event regardless.

This change introduces a new helper in `internal/trace`, `BackgroundContext`, which aliases `CopyContext(context.Background(), from)` with some more docstrings on when to use it and how to use it correctly. We use this new background context in the most inner part of `(TelemetryEventsExportQueueStore).QueueForExport` and `(EventLogStore).BulkInsert` to ensure we don't interrupt database inserts.

## Test plan

CI

### CODE REVIEW

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #57552** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57552" target="_blank">Graphite</a>  👈
    * **PR #57553** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57553" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57552?utm_source=stack-comment).

**Comment by bobheadxi** [comment]
I _want_ to backport this but I suppose it's not critical, so maybe we shouldn't 😅 cc @nathan-downs what do you think?

File: internal/trace/context.go [hunk 1]
```diff
@@ -22,6 +22,20 @@ func CopyContext(ctx context.Context, from context.Context) context.Context {
...
+// BackgroundContext returns a background context with the same tracing
+// policy and trace ID associated with from, if available. It is a convenience
+// alias for CopyContext(context.Background(), from).
+//
+// Using context.Background() is desirable in scenarios where we don't want an
+// action to be cancelled if the parent action is cancelled, while retaining a
+// trace hierarchy.
+//
+// 🚨 SECURITY: The returned context does NOT retain any other context baggage,
+// including Actor context.
```
Comment (camdencheek) [inline internal/trace/context.go:34]: drive-by: is this intentional? Because we can keep all the context values and just remove the cancellation with [context.WithoutCancel](https://pkg.go.dev/context#WithoutCancel)

Comment (bobheadxi) [reply internal/trace/context.go:34]: That's a big TIL

Comment (camdencheek) [reply internal/trace/context.go:34]: It's new 😄

Comment (bobheadxi) [reply internal/trace/context.go:34]: Ah - this is a 1.21 feature, and not available in Sourcegraph yet 😢

Comment (bobheadxi) [reply internal/trace/context.go:34]: I think I'll just copy the source code, it's quite simple https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/context/context.go;l=573-595

Comment (camdencheek) [reply internal/trace/context.go:34]: Oh, actually, it looks like it's only available as of go 1.21, which we're not using yet.

_However_, I created the utililty `(internal/xcontext).Detach` that does the same thing because I've needed this before. So as soon as we upgrade to 1.21, I'll replace all uses of `xcontext.Detach` with `context.WithoutCancel`

Comment (camdencheek) [reply internal/trace/context.go:34]: Hopefully we can upgrade to 1.21 soon! Just waiting on `rules_go` to get its ducks in a row 😄

Comment (bobheadxi) [reply internal/trace/context.go:34]: Actually I don't think that's possible, seems some of the secret sauce is that the implementation calls an internal function https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/context/context.go;l=754-785 with the wrapper context implementation. It seems this retrieves cancellation details from baggage as well, and there's special handling here for the wrapper

Comment (bobheadxi) [reply internal/trace/context.go]: Updated, thank you!

File: internal/database/event_logs.go [hunk 2]
```diff
@@ -290,7 +295,9 @@ func (l *eventLogStore) BulkInsert(ctx context.Context, events []*Event) error {
 	close(rowValues)
 
 	return batch.InsertValues(
-		ctx,
+		// Create a cancel-free context to avoid interrupting the insert when
+		// the parent context is cancelled.
+		xcontext.Detach(ctx),
```
Comment (camdencheek) [inline internal/database/event_logs.go:300]: nit: should we still have a timeout on here? I'm not confident this couldn't cause an indefinite hang if the remote never responds or something.

**Review by camdencheek — APPROVED** [review]

File: internal/database/event_logs.go (hunk 2 above)
Comment (bobheadxi) [reply internal/database/event_logs.go]: good point, I added a very high timeout of 5min, just to make sure it doesn't hang indefinitely. does that seem too high? There doesn't seem to be a similar case in the codebase that I can find

Comment (camdencheek) [reply internal/database/event_logs.go]: Seems fine to me!
